core/src/main/java/tachyon/underfs/UnderFileSystemRegistry.java

```diff
@@ -0,0 +1,256 @@
+ * priority. Factories registered this way will be registered at the start of the factories list so
+ * will have the first opportunity to indicate whether they support a requested path.
+ * </p>
+ */
+public class UnderFileSystemRegistry {
+
+  private static final List<UnderFileSystemFactory> FACTORIES =
+      new CopyOnWriteArrayList<UnderFileSystemFactory>();
+  private static final Logger LOG = LoggerFactory.getLogger(Constants.LOGGER_TYPE);
+  private static boolean sInit = false;
+
+  static {
+    // Call the actual initializer which is a synchronized method for thread safety purposes
+    init();
+  }
+
+  /**
+   * Returns a read-only view of the available factories
+   * 
+   * @return Read-only view of the available factories
+   */
+  public static List<UnderFileSystemFactory> available() {
+    return Collections.unmodifiableList(FACTORIES);
+  }
+
+  /**
+   * Creates a client that can talk to the under file system
+   * 
+   * @param path Path
+   * @param conf Optional configuration object for the UFS, may be null
+   * @param tachyonConf Tachyon Configuration
+   * @return Client for the under file system
+   * @throws IllegalArgumentException Thrown if there is no under file system for the given path or
+   *         if no under file system could successfully be created
+   */
+  public static UnderFileSystem create(String path, TachyonConf tachyonConf, Object conf) {
+    // Try to obtain the appropriate factory
+    List<UnderFileSystemFactory> factories = findAll(path, tachyonConf);
+    if (factories.size() == 0) {
+      throw new IllegalArgumentException("No known Under File System supports the given path "
+          + path);
+    }
+
+    List<Throwable> errors = new ArrayList<Throwable>();
+    for (UnderFileSystemFactory factory : factories) {
+      try {
+        // Use the factory to create the actual client for the Under File System
+        return factory.create(path, tachyonConf, conf);
+      } catch (Throwable e) {
+        errors.add(e);
+      }
+    }
+
+    // If we reach here no factories were able to successfully create for this path likely due to
+    // missing configuration since if we reached here at least some factories claimed to support the
+    // path
+    // Need to collate the errors
+    StringBuilder errorStr = new StringBuilder();
+    errorStr.append("All eligible Under File Systems were unable to create an instance for the "
+        + "given path: ").append(path).append('\n');
+    for (Throwable e: errors) {
+      errorStr.append(e.getMessage()).append('\n');
+    }
+    throw new IllegalArgumentException(errorStr.toString());
+  }
+
+  /**
+   * Finds the first Under File System factory that supports the given path
+   * 
+   * @param path Path
+   * @param conf Tachyon configuration
+   * @return Factory if available, null otherwise
+   */
+  public static UnderFileSystemFactory find(String path, TachyonConf conf) {
```

Comment:
why doesn't this have 'Object conf' parameter. Please create a JIRA to investigate this along with the last one.